File: 链表/25.K个一组翻转链表.go
// Go 版本
// Definition:
//
//	type ListNode struct {
//	    Val  int
//	    Next *ListNode
//	}
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || k <= 1 {
		return head
	}
	dummy := &ListNode{Val: -1, Next: head}
	pre := dummy

	for {
		// 找到当前组的尾节点 tail，若不足 k 个则返回结果
		tail := pre
		for i := 0; i < k; i++ {
			tail = tail.Next
			if tail == nil {
				return dummy.Next
			}
		}
		nextGroupHead := tail.Next

		// 原地反转 [pre.Next, tail]（使用 prev 指向 nextGroupHead 便于尾部连接）
		prev := nextGroupHead
		cur := pre.Next
		for cur != nextGroupHead {
			tmp := cur.Next
			cur.Next = prev
			prev = cur
			cur = tmp
		}

		// 接回反转后的部分，并移动 pre 到已处理段的尾部
		oldHead := pre.Next
		pre.Next = prev
		pre = oldHead
	}
}

File: 链表/25.K个一组翻转链表.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head==null || k<=1) return head;
        ListNode dummy = new ListNode(-1,head);
        ListNode pre = dummy;
        while(true){
            ListNode tail = pre;
            for(int i=0;i<k;i++){
                tail=tail.next;
                if(tail==null){
                    return dummy.next;
                }
            }
            ListNode nextGroupHead = tail.next;

            ListNode prev = nextGroupHead;
            ListNode cur = pre.next;
            while(cur!=nextGroupHead){
                ListNode tmp = cur.next;
                cur.next=prev;
                prev = cur;
                cur = tmp;
            }
            ListNode oldHead = pre.next;
            pre.next=prev;
            pre = oldHead;
        }
    }
}